북마크 기능 수정

june0122 opened the pull request:
#### BookmarkFragment

```kotlin
// 북마크 화면에서는 `isBookmark` 상태 체크 필요없이 바로 북마크 버튼 이미지 활성화
viewModel.bookmarks.observe(viewLifecycleOwner) { bookmarks ->
    bookmarks.forEachIndexed { position, user ->
        // 북마크 화면 진입 시에는 holder가 초기화되어 있지 않으므로 bookmarks에 변경이 발생해야 북마크 버튼 이미지가 활성화 됨
        val holder = recyclerView.findViewHolderForAdapterPosition(position)
        val bookmarkButton = holder?.itemView?.findViewById<ImageView>(R.id.btn_bookmark)
        bookmarkButton?.setImageResource(R.drawable.ic_bookmark_filled)
    }
    bookmarkAdapter.updateUserListItems(bookmarks)
}
```

#### UserListFragment

```kotlin
viewModel.items.observe(requireActivity()) { users ->
    users.forEachIndexed { position, user ->
        val holder = recyclerView.findViewHolderForAdapterPosition(position)
        val bookmarkButton = holder?.itemView?.findViewById<ImageView>(R.id.btn_bookmark)
        if (user is User && user.isBookmark) {
            bookmarkButton?.setImageResource(R.drawable.ic_bookmark_filled)
        } else {
            bookmarkButton?.setImageResource(R.drawable.ic_bookmark)
        }
    }
    userListAdapter.updateUserListItems(users)
}
```

june0122 committed:
refactor: RepositoryModule에서 Retrofit 관련 코드 NetworkModule로 분리

june0122 committed:
refactor: suspend 함수에 Dispatchers 정의

june0122 committed:
fix: 뷰모델에서 레파지토리에 Flow로 정의된 bookmarks 데이터 동기화

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/main/UserSharedViewModel.kt:
```diff
@@ -129,10 +131,12 @@ class UserSharedViewModel @Inject constructor(private val repository: UserReposi
 
     fun delete(user: User) = viewModelScope.launch {
         repository.delete(user)
+        repository.allUsers.onEach { checkBookmarks(it) }.launchIn(viewModelScope)
```
flow를 활용할대는 매번 launchIn을 하는게 아니라, 딱 하나만 만드시면 됩니다.
지금 insert, delete를 N 번하면 새로운 observe가 N번 일어나요.
지난번에 했던 livedata 와 동일합니다.

june0122 committed:
refactor: 북마크 버튼의 리소스 변경을 뷰홀더에 접근하지 않고 아닌 뷰홀더 내에서 처리하도록 변경

june0122 committed:
fix: `isBookmark` Boolean 값과 북마크의 이미지가 일치 안되는 버그 수정

june0122 committed:
refactor: 메서드 이름 변경 및 주석 삭제

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/data/room/UserDao.kt:
```diff
@@ -9,11 +12,14 @@ interface UserDao {
-    @Delete
-    suspend fun delete(user: User)
+    @Query("DELETE FROM user_table WHERE name LIKE (:name)")
+    suspend fun delete(name: String)
```
@Delete를 이용하는것과 @Query를 이용한것의 차이가 뭘까요?

june0122 replied:
`@query` 어노테이션을 사용하면 일반적인 `@delete`, `@insert` 등을 사용할 때보다 상세한 조건을 추가하여 삽입 또는 삭제를 할 수 있다고 이해하고 있습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/main/UserSharedViewModel.kt:
```diff
@@ -22,6 +25,12 @@ import javax.inject.Inject
 
+    init {
+        viewModelScope.launch {
+            repository.allUsers.onEach { checkBookmarks(it) }.launchIn(viewModelScope)
```
onEach는 어땔때 사용할까요?

그리고 이 코드는 viewModelScope.launch {}를 하실 필요가 없습니다.
launchIn만으로도 이 코드는 동작하는데, 왜 그럴까요?

june0122 replied:
`onEach`는 repository 내부의 데이터에 이벤트가 발생했을 때 리스너처럼 특정 동작을 정의할 때 사용할 수 있는 것으로 알고 있습니다.

`viewModelScope.launch {}` 사용하지 않아도 `launchIn` 함수 자체가 `scope.lauch { flow.collect() }`의 역할을 하는 코드이기 때문에 스코프를 별도로 지정해주지 않아도 됩니다!

june0122 committed:
refactor: `launchIn` 함수의 역할과 중복되는 스코프 지정 제거

viewModelScope.launch {} 사용하지 않아도 launchIn 함수 자체가
scope.lauch { flow.collect() }의 역할을 하는 코드이기 때문에
스코프를 별도로 지정해주지 않아도 된다.
